File: src/main/java/frc/chargers/hardware/motorcontrol/ChargerSpark.java
package frc.chargers.hardware.motorcontrol;

import com.revrobotics.AbsoluteEncoder;
import com.revrobotics.RelativeEncoder;
import com.revrobotics.spark.SparkBase;
import com.revrobotics.spark.SparkClosedLoopController;
import com.revrobotics.spark.SparkFlex;
import com.revrobotics.spark.SparkMax;
import com.revrobotics.spark.SparkSim;
import com.revrobotics.spark.config.ClosedLoopConfig;
import com.revrobotics.spark.config.ClosedLoopConfig.FeedbackSensor;
import com.revrobotics.spark.config.SparkBaseConfig;
import com.revrobotics.spark.config.SparkFlexConfig;
import com.revrobotics.spark.config.SparkMaxConfig;
import edu.wpi.first.hal.HAL;
import edu.wpi.first.math.MathUtil;
import edu.wpi.first.math.system.plant.DCMotor;
import edu.wpi.first.units.measure.Angle;
import edu.wpi.first.wpilibj.Alert;
import edu.wpi.first.wpilibj.Alert.AlertType;
import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.RobotBase;
import edu.wpi.first.wpilibj.RobotController;
import frc.chargers.hardware.encoders.Encoder;
import frc.chargers.utils.Tracer;
import org.jetbrains.annotations.Nullable;

import java.util.Optional;

import static com.revrobotics.spark.ClosedLoopSlot.kSlot0;
import static com.revrobotics.spark.ClosedLoopSlot.kSlot1;
import static com.revrobotics.spark.SparkBase.PersistMode.kPersistParameters;
import static com.revrobotics.spark.SparkBase.ResetMode.kNoResetSafeParameters;
import static com.revrobotics.spark.SparkBase.ResetMode.kResetSafeParameters;
import static com.revrobotics.spark.SparkLowLevel.MotorType;
import static edu.wpi.first.math.util.Units.*;
import static edu.wpi.first.units.Units.Rotations;
import static frc.chargers.utils.UtilMethods.tryUntilOk;
import static java.lang.Math.PI;

/**
 * A Spark Max/Spark Flex that implements the Motor interface.
 * To access more low-level capabilities of the base api, inherit from this class.
 */
public class ChargerSpark implements Motor {
	protected final SparkBase baseApi;
	protected RelativeEncoder relativeEncoder;
	protected final SparkClosedLoopController pidController;
	protected SparkBaseConfig initialConfig;
	protected Encoder encoder = new Encoder() {
		@Override
		public double positionRad() {
			return rotationsToRadians(relativeEncoder.getPosition());
		}
		
		@Override
		public double velocityRadPerSec() {
			return rotationsPerMinuteToRadiansPerSecond(relativeEncoder.getVelocity());
		}
		
		@Override
		public void setPositionReading(Angle angle) {
			relativeEncoder.setPosition(angle.in(Rotations));
		}
	};
	
	public enum Model {
		SPARK_MAX, SPARK_FLEX
	}
	
	public ChargerSpark(int id, Model model, @Nullable SparkBaseConfig config) {
		this.baseApi = model == Model.SPARK_MAX ? new SparkMax(id, MotorType.kBrushless) : new SparkFlex(id, MotorType.kBrushless);
		this.relativeEncoder = baseApi.getEncoder();
		this.pidController = baseApi.getClosedLoopController();
		if (config != null) {
			this.initialConfig = config;
			tryUntilOk(baseApi, () -> baseApi.configure(config, kResetSafeParameters, kPersistParameters));
		} else {
			this.initialConfig = model == Model.SPARK_MAX ? new SparkMaxConfig() : new SparkFlexConfig();
		}
	}
	
	public ChargerSpark withAbsoluteEncoder() {
		var absoluteEncoder = baseApi.getAbsoluteEncoder();
		initialConfig.closedLoop.feedbackSensor(FeedbackSensor.kAbsoluteEncoder);
		tryUntilOk(baseApi, () -> baseApi.configure(initialConfig, kResetSafeParameters, kPersistParameters));
		encoder = new Encoder() {
			@Override
			public double positionRad() {
				return rotationsToRadians(absoluteEncoder.getPosition());
			}
			
			@Override
			public double velocityRadPerSec() {
				return rotationsToRadians(absoluteEncoder.getVelocity());
			}
			
			@Override
			public void setPositionReading(Angle angle) {
				DriverStation.reportError("Position reading cannot be set for absolute encoder.", false);
			}
		};
		return this;
	}
	
	public ChargerSpark withSim(SimDynamics dynamics, DCMotor motorType) {
		if (!RobotBase.isSimulation()) return this;
		var sim = new SparkSim(baseApi, motorType);
		HAL.registerSimPeriodicAfterCallback(() -> {
			double batteryVoltage = RobotController.getBatteryVoltage();
			dynamics.acceptVolts().accept(MathUtil.clamp(sim.getAppliedOutput() * batteryVoltage, -12, 12));
			sim.iterate(radiansPerSecondToRotationsPerMinute(dynamics.velocity().getAsDouble()), batteryVoltage, 0.02);
		});
		return this;
	}
	
	public ChargerSpark enableAlternateEncoder() {
		if (baseApi instanceof SparkMax m) {
			relativeEncoder = m.getAlternateEncoder();
		} else if (baseApi instanceof SparkFlex m) {
			relativeEncoder = m.getExternalEncoder();
		} else {
			new Alert("Invalid spark max type -- code issue lmao", AlertType.kError).set(true);
		}
		return this;
	}
	
	@Override
	public Encoder encoder() { return encoder; }
	
	@Override
	public double outputVoltage() {
		return baseApi.getAppliedOutput() * baseApi.getBusVoltage();
	}
	
	@Override
	public double statorCurrent() { return baseApi.getOutputCurrent(); }
	
	@Override
	public double tempCelsius() { return baseApi.getMotorTemperature(); }
	
	@Override
	public int id() {
		return baseApi.getDeviceId();
	}
	
	@Override
	public void setVoltage(double volts) { baseApi.setVoltage(volts); }
	
	@Override
	public void setVelocity(double velocityRadPerSec, double ffVolts) {
		pidController.setReference(
			radiansToRotations(velocityRadPerSec),
			SparkBase.ControlType.kVelocity,
			kSlot1, ffVolts
		);
	}
	
	@Override
	public void moveToPosition(double positionRads, double ffVolts) {
		pidController.setReference(
			radiansToRotations(positionRads),
			SparkBase.ControlType.kPosition,
			kSlot0, ffVolts
		);
	}
	
	@Override
	public void setCoastMode(boolean enabled) {
		Tracer.startTrace("set coast mode(spark)");
		baseApi.configure(
			initialConfig.idleMode(enabled ? SparkBaseConfig.IdleMode.kCoast : SparkBaseConfig.IdleMode.kBrake),
			kNoResetSafeParameters,
			kPersistParameters
		);
		Tracer.endTrace();
	}
	
	@Override
	public void setControlsConfig(ControlsConfig newConfig) {
		if (newConfig.positionPID().kP != 0.0) {
			initialConfig.closedLoop
				.pid(newConfig.positionPID().kP * (2 * PI), newConfig.positionPID().kI * (2 * PI), newConfig.positionPID().kD * (2 * PI), kSlot0)
				.positionWrappingEnabled(newConfig.continuousInput())
				.positionWrappingInputRange(-PI, PI);
		}
		if (newConfig.velocityPID().kP != 0.0) {
			initialConfig.closedLoop
				.pid(newConfig.velocityPID().kP * (2 * PI), newConfig.velocityPID().kI * (2 * PI), newConfig.velocityPID().kD * (2 * PI), kSlot1);
		}
		if (newConfig.gearRatio() != 1.0) {
			initialConfig.encoder
				.positionConversionFactor(1 / newConfig.gearRatio())
				.velocityConversionFactor(1 / newConfig.gearRatio());
		}
		tryUntilOk(baseApi, () -> baseApi.configure(initialConfig, kResetSafeParameters, kPersistParameters));
	}
	
	@Override
	public void close() {
		if (RobotBase.isSimulation()) baseApi.close();
	}
}

File: src/main/kotlin/frc/chargers/hardware/motorcontrol/ChargerSpark.kt
package frc.chargers.hardware.motorcontrol

import com.batterystaple.kmeasure.quantities.*
import com.batterystaple.kmeasure.units.*
import com.pathplanner.lib.util.PIDConstants
import com.revrobotics.*
import com.revrobotics.CANSparkLowLevel.PeriodicFrame
import edu.wpi.first.wpilibj.Alert
import edu.wpi.first.wpilibj.Alert.AlertType
import frc.chargers.framework.UnitTesting
import frc.chargers.hardware.encoders.Encoder
import frc.chargers.utils.units.frequencyToPeriod
import kotlin.math.PI
import kotlin.math.roundToInt


/**
 * A [CANSparkMax] motor controller that implements the [Motor] interface
 * and performs periodic self-checking.
 *
 * The [base] property allows for the access of the base [CANSparkMax] instance.
 */
class ChargerSparkMax(
    deviceID: Int,
    gearRatio: Double,
    motorType: CANSparkLowLevel.MotorType = CANSparkLowLevel.MotorType.kBrushless,
    useAbsoluteEncoder: Boolean = false,
    factoryDefault: Boolean = true,
    faultLogName: String? = null
): ChargerSpark<CANSparkMax>(
    CANSparkMax(deviceID, motorType),
    gearRatio, useAbsoluteEncoder, factoryDefault, faultLogName
)


/**
 * A [CANSparkFlex] motor controller that implements the [Motor] interface
 * and performs periodic self-checking.
 *
 * The [base] property allows for the access of the base [CANSparkFlex] instance.
 */
class ChargerSparkFlex(
    deviceID: Int,
    gearRatio: Double,
    useAbsoluteEncoder: Boolean = false,
    factoryDefault: Boolean = true,
    faultLogName: String? = null
): ChargerSpark<CANSparkFlex>(
    CANSparkFlex(deviceID, CANSparkLowLevel.MotorType.kBrushless),
    gearRatio, useAbsoluteEncoder, factoryDefault, faultLogName
)


open class ChargerSpark<BaseMotorType: CANSparkBase>(
    /**
     * The base [CANSparkFlex]/[CANSparkMax] instance.
     */
    val base: BaseMotorType,
    gearRatio: Double,
    private val useAbsoluteEncoder: Boolean = false,
    factoryDefault: Boolean = true,
    private val faultLogName: String? = null
): Motor {
    val deviceID: Int = base.deviceId

    private val nonRevFollowers = mutableListOf<Motor>()
    private val relativeEncoder = base.encoder
    private val absoluteEncoder = base.getAbsoluteEncoder(SparkAbsoluteEncoder.Type.kDutyCycle)

    private var positionPIDConfigured = false
    private var velocityPIDConfigured = false

    private val errors = mutableListOf<REVLibError>()
    private fun REVLibError.bind() { if (this != REVLibError.kOk) errors.add(this) }
    private val faultAlert = Alert("Spark Max Error", "", AlertType.kError)

    init {
        UnitTesting.addGlobalCloseable(base)
        for (i in 1..4) {
            errors.clear()
            if (factoryDefault) base.restoreFactoryDefaults().bind()
            if (useAbsoluteEncoder) {
                base.pidController.setFeedbackDevice(absoluteEncoder).bind()
                absoluteEncoder.setPositionConversionFactor(1 / gearRatio).bind()
                absoluteEncoder.setVelocityConversionFactor(1 / gearRatio).bind()
            } else {
                base.pidController.setFeedbackDevice(relativeEncoder).bind()
                relativeEncoder.setPositionConversionFactor(1 / gearRatio).bind()
                relativeEncoder.setVelocityConversionFactor(1 / gearRatio).bind()
            }
            base.burnFlash().bind()
            if (errors.isEmpty()) break
        }
        require(errors.isEmpty()) { "Gear Ratio of ChargerSparkMax(id=$deviceID) not configured; critical error." }
    }

    override fun logFaults() {
        val err = base.lastError
        if (err != REVLibError.kOk) {
            faultAlert.setText(err.toString())
            faultAlert.set(true)
        }
    }

    override val encoder: Encoder = if (useAbsoluteEncoder) AbsoluteEncoderImpl() else RelativeEncoderImpl()
    private inner class RelativeEncoderImpl: Encoder {
        override val angularVelocity: AngularVelocity
            get() = relativeEncoder.velocity.ofUnit(rotations / minutes)

        override val angularPosition: Angle
            get() = relativeEncoder.position.ofUnit(rotations)
    }
    private inner class AbsoluteEncoderImpl: Encoder {
        override val angularVelocity: AngularVelocity
            get() = absoluteEncoder.velocity.ofUnit(rotations / seconds)

        override val angularPosition: Angle
            get() = absoluteEncoder.position.ofUnit(rotations)
    }

    override var voltageOut: Voltage
        get() = (base.appliedOutput * base.busVoltage).ofUnit(volts)
        set(voltage){
            base.setVoltage(voltage.inUnit(volts))
        }

    override val statorCurrent get() = base.outputCurrent.ofUnit(amps)

    override val inverted get() = base.inverted

    override fun setPositionSetpoint(position: Angle, feedforward: Voltage) {
        if (!positionPIDConfigured) {
            Motor.alertVelocityPIDErr()
            return
        } else if (abs(position - this.encoder.angularPosition) < 1.degrees) {
            base.setVoltage(0.0)
            return
        }
        base.pidController.setReference(
            position.inUnit(rotations),
            CANSparkBase.ControlType.kPosition,
            0,
            feedforward.inUnit(volts)
        )
    }

    override fun setVelocitySetpoint(velocity: AngularVelocity, feedforward: Voltage) {
        if (!velocityPIDConfigured) {
            Motor.alertVelocityPIDErr()
            return
        }
        base.pidController.setReference(
            if (useAbsoluteEncoder) {
                velocity.inUnit(rotations / seconds)
            } else {
                velocity.inUnit(rotations / minutes)
            },
            CANSparkBase.ControlType.kVelocity,
            1,
            feedforward.inUnit(volts)
        )
    }

    override fun configure(
        inverted: Boolean?,
        brakeWhenIdle: Boolean?,
        rampRate: Time?,
        statorCurrentLimit: Current?,
        followerMotors: List<Motor>,
        positionUpdateRate: Frequency?,
        velocityUpdateRate: Frequency?,
        optimizeUpdateRate: Boolean?,
        currentPosition: Angle?,
        positionPID: PIDConstants?,
        velocityPID: PIDConstants?,
        continuousInput: Boolean?
    ): ChargerSpark<BaseMotorType> {
        if (inverted != null) base.inverted = inverted
        for (i in 1..4) {
            errors.clear()
            when (brakeWhenIdle) {
                true -> base.setIdleMode(CANSparkBase.IdleMode.kBrake).bind()
                false -> base.setIdleMode(CANSparkBase.IdleMode.kCoast).bind()
                null -> {}
            }
            if (rampRate != null) {
                base.setOpenLoopRampRate(rampRate.inUnit(seconds)).bind()
                base.setClosedLoopRampRate(rampRate.inUnit(seconds)).bind()
            }
            if (statorCurrentLimit != null) base.setSmartCurrentLimit(statorCurrent.inUnit(amps).roundToInt()).bind()
            for (follower in followerMotors) {
                follower.configure(
                    positionPID = positionPID,
                    velocityPID = velocityPID,
                    currentPosition = currentPosition
                )
                when (follower) {
                    is ChargerSpark<*> -> follower.base.follow(base, follower.inverted).bind()
                    else -> nonRevFollowers.add(follower)
                }
            }
            if (currentPosition != null) relativeEncoder.setPosition(currentPosition.inUnit(rotations)).bind()
            // 2 * PI makes it so that the PID gains are optimized off of radians and not rotations
            if (positionPID != null) {
                positionPIDConfigured = true
                base.pidController.apply {
                    setP(positionPID.kP * (2 * PI), 0).bind()
                    setI(positionPID.kI * (2 * PI), 0).bind()
                    setD(positionPID.kD * (2 * PI), 0).bind()
                }
            }
            if (velocityPID != null) {
                velocityPIDConfigured = true
                // relative encoder pid is based off of rotations/min so we have to compensate here
                val multiplier = 2 * PI * if (useAbsoluteEncoder) 1.0 else 60.0
                base.pidController.apply {
                    setP(velocityPID.kP * multiplier, 1).bind()
                    setI(velocityPID.kI * multiplier, 1).bind()
                    setD(velocityPID.kD * multiplier, 1).bind()
                }
            }
            if (continuousInput == true) {
                base.pidController.apply {
                    setPositionPIDWrappingEnabled(true).bind()
                    setPositionPIDWrappingMinInput(-180.degrees.inUnit(rotations)).bind()
                    setPositionPIDWrappingMaxInput(180.degrees.inUnit(rotations)).bind()
                }
            } else if (continuousInput == false) {
                base.pidController.setPositionPIDWrappingEnabled(false).bind()
            }
            if (optimizeUpdateRate == true) {
                val disabledFrames = if (useAbsoluteEncoder) {
                    listOf(PeriodicFrame.kStatus2, PeriodicFrame.kStatus3, PeriodicFrame.kStatus4)
                } else {
                    listOf(PeriodicFrame.kStatus3, PeriodicFrame.kStatus4, PeriodicFrame.kStatus5, PeriodicFrame.kStatus6)
                }
                for (frame in disabledFrames) {
                    base.setPeriodicFramePeriod(frame, 65535)
                }
            }
            if (positionUpdateRate != null) {
                base.setPeriodicFramePeriod(
                    if (useAbsoluteEncoder) PeriodicFrame.kStatus5 else PeriodicFrame.kStatus2,
                    frequencyToPeriod(positionUpdateRate).inUnit(milli.seconds).roundToInt()
                ).bind()
            }
            if (velocityUpdateRate != null) {
                base.setPeriodicFramePeriod(
                    if (useAbsoluteEncoder) PeriodicFrame.kStatus6 else PeriodicFrame.kStatus1,
                    frequencyToPeriod(velocityUpdateRate).inUnit(milli.seconds).roundToInt()
                ).bind()
            }
            base.burnFlash().bind()
            if (errors.isEmpty()) return this
        }
        Alert(
            "${faultLogName ?: "ChargerSpark($deviceID)"} could not configure. Errors: $errors",
            AlertType.kError
        ).set(true)
        return this
    }
}